### ISSUE TITLE
About optimization-based bounding box refinement

### ISSUE THREAD

**Issue by sty-yyj** [vacancy/PreciseRoIPooling#22]
Hi, thanks for your great work! I've been very interested in optimization-based bounding box refinement! But I have a question about it.
The bbox is in the form of  (x_1, y_1, x_2, y_2), its coordinate values only participate in PrRoiPooling calculation part, so how to update bbox via back propagation.
I try to make it clear by implementing it on Pytroch, but the *(grad, b_j)* I got is **None**!
Could you tell me more details about this?
Thank you!

**Comment by vacancy** [comment]
Make sure you have `bbox_coordinates.requires_grad = True` See also:
https://github.com/vacancy/PreciseRoIPooling/blob/master/pytorch/tests/test_prroi_pooling2d.py#L37-L52

**Comment by sty-yyj** [comment]
Thank you for your reply. I've solved this question about box refinement!
But I still have two problems while I try to implement IoUNet.
- **IoU label in IoU predictor branch**. According to your paper, *IoU labels are normalized s.t. the values are distributed over [−1, 1]*. So I normalize IoU labels by $IoU * 2 - 1$. For example, if a box's true IoU is `0.8`, `classes_number` is `5` and this box is belong to first category,  its **IoU label** should be `[0.6, -1, -1, -1, -1]`. Is it right? 
- **Outputs in IoU predictor branch**. I use `tanh()` function to limit my IoU predictor outputs in [-1, 1]. But the result of experiment show that **all outputs of IoU predictor tend to be -1**, it seems caused by vanishing gradient.

Could you offer me some suggestions?
Sorry to bother again!

**Comment by vacancy** [comment]
The label is [0.6, -1, -1, -1, -1]. However, you need an extra mask tensor to ignore the -1's (they should not back propagate). See also how typical bounding box regression heads are trained.
For example, https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/maskrcnn_benchmark/modeling/roi_heads/box_head/loss.py#L151-L165

**Comment by sty-yyj** [comment]
Actually I‘ve tried this，but the result is not very well.
So I suspect that using `tanh()` normalization is questionable.
Could you tell me how do you limit the outputs of IoU predictor branch to [-1, 1]?

Thank you for your reply again!

**Comment by vacancy** [comment]
The output branch is Tanh-activated.

Implementation details from the paper:
> We generate bounding boxes and labels for training the IoU-Net by augmenting the ground- truth, instead of taking proposals from RPNs. Specifically, for all ground-truth bounding boxes in the training set, we manually transform them with a set of randomized parameters, resulting in a candidate bounding box set. **_We then remove from this candidate set the bounding boxes having an IoU less than Ωtrain = 0.5 with the matched ground-truth._** **_We uniformly sample training data from this candidate set w.r.t. the IoU._** 

> We use smooth-L1 loss for training the IoU predictor. The training data for the IoU predictor is separately generated as described in Section 3.1 within images in a training batch. **_IoU labels are normalized s.t. the values are distributed over [−1, 1]._**

1. Since there is no bbox with IoU >= 0.5, the label normalization equation should be 4 * IoU - 3.
2. You should uniformly sample training boxes w.r.t. the IoU label.

**Comment by sty-yyj** [comment]
That's very helpful!
Thank you!

**Comment by AlexanderHustinx** [comment]
Hey @sty-yyj,

You mentioned that you managed to implement the optimisation-based bounding box refinement?
I'm having some trouble myself, would you mind helping me out?
